### ISSUE TITLE
Add startup and liveness probe field for Cloud Run service resource

### ISSUE THREAD

**Issue by yanweiguo** [hashicorp/terraform-provider-google#12532]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment. If the issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If the issue is assigned to a user, that user is claiming responsibility for the issue. If the issue is assigned to "hashibot", a community member has claimed the issue already.

### Description

Adds new fields `startup_probe` and `liveness_probe` to google cloud run service resource.

### New or Affected Resource(s)

* google_cloud_run_service

### Potential Terraform Configuration

```tf
resource "google_cloud_run_service" "service" {
  name     = "my-service"
  location = "us-central1"
  metadata {
    annotations = {
      "run.googleapis.com/launch-stage" = "BETA"
    }
  }

  template {
    spec {
      containers {
        image = "us-docker.pkg.dev/cloudrun/container/hello"
        startup_probe {
          initial_delay_seconds = 0
          timeout_seconds = 1
          period_seconds = 3
          failure_threshold = 1
          tcpSocket {
            port = 8080
          }
        }
      }
    }
  }

  traffic {
    percent         = 100
    latest_revision = true
  }
}
```

### References

* https://cloud.google.com/run/docs/reference/rest/v1/Container

**Comment by marcusricardoaguiar** [comment]
It would also be good to include the liveness_probe. :)

**Comment by yanweiguo** [comment]
> It would also be good to include the liveness_probe. :)

Changed the title :)

**Comment by darkangel01111991** [comment]
Would love to get this.

Have you found any workaround to setup a startup/liveness probe while avoiding the misalignment between terraform state and the current infra?

Any short term solution would be much appreciated! Thanks a lot!

**Comment by Thomasdezeeuw** [comment]
Not sure if this is the correct place to report bugs or if I should open a new issue instead. I noticed that `startup_probe` is not removed (properly) once added. Steps to reproduces:
1. Create a Cloud Run service with `startup_probe` (plan & apply).
2. Remove the `startup_probe` config.
3. Run `terraform apply` and you'll notice that it doesn't actually remove the `startup_probe` related configuration.

**Comment by yanweiguo** [comment]
> Not sure if this is the correct place to report bugs or if I should open a new issue instead. I noticed that `startup_probe` is not removed (properly) once added. Steps to reproduces:
> 
> 1. Create a Cloud Run service with `startup_probe` (plan & apply).
> 2. Remove the `startup_probe` config.
> 3. Run `terraform apply` and you'll notice that it doesn't actually remove the `startup_probe` related configuration.

It's intended behavior as the result of that this field is optional and it has a default value. You can't actually get rid of startup probe.

Before Cloud Run supports a customized startup probe, it always applies a TCP health check for a Cloud Run service deployment although it is not reflected in the API. Cloud Run will capture this default TCP startup probe in the API soon.

**Comment by Thomasdezeeuw** [comment]
> It's intended behavior as the result of that this field is optional and it has a default value. You can't actually get rid of startup probe.
> 
> Before Cloud Run supports a customized startup probe, it always applies a TCP health check for a Cloud Run service deployment although it is not reflected in the API. Cloud Run will capture this default TCP startup probe in the API soon.

I didn't know about this. Is this something worth calling out explicitly in the documentation? I found it a rather annoying gotcha when debugging a startup probe issue.

**Comment by yanweiguo** [comment]
> Probes are now GA (According to https://cloud.google.com/run/docs/release-notes) Does that make it safe to remove the beta annotations?

Yes. Beta launch stage annotation is no longer required.

**Comment by cagataygurturk** [comment]
This issue should be closed maybe? I have opened #13310 to move the feature to stable provider though.

**Comment by yanweiguo** [comment]
> This issue should be closed maybe? I have opened #13310 to move the feature to stable provider though.

Thanks for opening that one.
